packages/core-data/src/selectors-ts.ts

```diff
@@ -0,0 +1,140 @@
+/**
+ * External dependencies
+ */
+import createSelector from 'rememo';
+import { set, get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { getNormalizedCommaSeparable } from './utils';
+import type {
+	Context,
+	EntityQuery,
+	KeyOf,
+	Kind,
+	KindOf,
+	Name,
+	NameOf,
+	RecordOf,
+	DefaultContextOf,
+} from './types';
+
+// Placeholder State type for now
+export type State = any;
+
+/**
+ * Returns the Entity's record object by key. Returns `null` if the value is not
+ * yet received, undefined if the value entity is known to not exist, or the
+ * entity object if it exists and is received.
+ *
+ * @param  state State tree
+ * @param  kind  Entity kind.
+ * @param  name  Entity name.
+ * @param  key   Record's key
+ * @param  query Optional query.
+ *
+ * @return Record.
+ */
+export const getEntityRecord = createSelector(
+	function <
+		R extends RecordOf< K, N >,
+		C extends Context = DefaultContextOf< R >,
+		K extends Kind = KindOf< R >,
+		N extends Name = NameOf< R >,
+		Q extends EntityQuery< any > = EntityQuery< C >
+	>(
+		state: State,
+		kind: K,
+		name: N,
+		key: KeyOf< R >,
+		query?: Q
+	):
+		| ( Q[ '_fields' ] extends string[]
+				? Partial< RecordOf< K, N, C > >
+				: RecordOf< K, N, C > )
+		| null
+		| undefined {
```

Comment:
is `null` vs. `undefined` important? would it be better if we only returned one of those values?

Reply:
There are some [empty test cases](https://github.com/WordPress/gutenberg/blob/8fbf70c6b6aa965af4eae4bb052b7a86acbdc1b9/packages/core-data/src/test/selectors.js#L28) that indicate the `null` is important, which makes me think it's part of the de-facto public API signature of this function.

That being said, I'd love if it just returned one or another. Let's proceed with that `null` here to avoid getting blocked, and take this discussion to another PR.